src/SdkCommon/ClientRuntime/ClientRuntime/Serialization/ExtensibleEnumConverter.cs

```diff
@@ -0,0 +1,222 @@
+                        expectedValue,
+                        reader.Value);
+
+                throw new JsonSerializationException(message);
+            }
+
+            TValue result = default(TValue);
+
+            if (reader.Value != null)
+            {
+                if (!typeof(TValue).GetTypeInfo().IsAssignableFrom(reader.ValueType.GetTypeInfo()))
+                {
+                    string message =
+                        String.Format(
+                            "Deserialization failed. Value '{0}' is not of expected type '{1}'.",
+                            reader.Value,
+                            typeof(TValue));
+
+                    throw new JsonSerializationException(message);
+                }
+
+                result = (TValue)reader.Value;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Advances the given JSON reader, or throws a JsonSerializationException if it cannot be advanced.
+        /// </summary>
+        /// <param name="reader">The JSON reader to advance.</param>
+        protected void Advance(JsonReader reader)
+        {
+            if (!reader.Read())
+            {
+                throw new JsonSerializationException("Deserialization failed. Unexpected end of input.");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Serializes and deserializes "extensible enums" to and from JSON. Extensible enums are like enumerations in
+    /// that they have well-known values, but they are extensible with new values and the values are based on strings
+    /// instead of integers.
+    /// </summary>
+    public class ExtensibleEnumConverter<T> : ConverterBase where T : ExtensibleEnum<T>
+    {
+        /// <summary>
+        /// Delegate type for a factory method that creates or looks up an ExtensibleEnum instance from a given string.
+        /// </summary>
+        /// <typeparam name="T">The type of ExtensibleEnum returned.</typeparam>
+        /// <param name="name">The enum value to look up or create.</param>
+        /// <returns>An instance of type T.</returns>
+        public delegate T ExtensibleEnumValueFactory<T>(string name) where T : ExtensibleEnum<T>;
+
+        private ExtensibleEnumValueFactory<T> _enumValueFactory;
+
+        /// <summary>
+        /// Initializes a new instance of the ExtensibleEnumConverter class.
+        /// </summary>
+        public ExtensibleEnumConverter() : this("Create") { }
+
+        /// <summary>
+        /// Initializes a new instance of the ExtensibleEnumConverter class.
+        /// </summary>
+        /// <param name="factoryMethodName">
+        /// The name of a public static method that creates an instance of type T given a string value; Default is
+        /// "Create".
+        /// </param>
+        public ExtensibleEnumConverter(string factoryMethodName)
+        {
+            if (factoryMethodName == null) {
```

Comment:
nit: Inconsistent brace indentation.

Unsolicited sales pitch: Consider adopting the `Throw` class. It makes code like this so much shorter, and FxCop is still smart enough to see that argument checking is happening.

Reply:
Introducing newer types here may raise some concerns for @shahabhijeet  I shall leave it to his discretion